network/p2p/libp2pStream_test.go

```diff
@@ -0,0 +1,457 @@
+package p2p
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	swarm "github.com/libp2p/go-libp2p-swarm"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// TestStreamClosing tests 1-1 communication with streams closed using libp2p2 handler.FullClose
+func TestStreamClosing(t *testing.T) {
+	count := 10
+	ctx, cancel := context.WithCancel(context.Background())
+	var msgRegex = regexp.MustCompile("^hello[0-9]")
+
+	handler, streamCloseWG := mockStreamHandlerForMessages(t, ctx, count, msgRegex)
+
+	// Creates nodes
+	nodes, identities := NodesFixtureWithHandler(t, 2, handler, false)
+	defer StopNodes(t, nodes)
+	defer cancel()
+
+	nodeInfo1, err := PeerAddressInfo(*identities[1])
+	require.NoError(t, err)
+
+	senderWG := sync.WaitGroup{}
+	senderWG.Add(count)
+	for i := 0; i < count; i++ {
+		go func(i int) {
+			// Create stream from node 1 to node 2 (reuse if one already exists)
+			nodes[0].host.Peerstore().AddAddrs(nodeInfo1.ID, nodeInfo1.Addrs, peerstore.AddressTTL)
+			s, err := nodes[0].CreateStream(context.Background(), nodeInfo1.ID)
+			assert.NoError(t, err)
+			w := bufio.NewWriter(s)
+
+			// Send message from node 1 to 2
+			msg := fmt.Sprintf("hello%d\n", i)
+			_, err = w.WriteString(msg)
+			assert.NoError(t, err)
+
+			// Flush the stream
+			assert.NoError(t, w.Flush())
+
+			// close the stream
+			err = s.Close()
+			require.NoError(t, err)
+
+			senderWG.Done()
+		}(i)
+	}
+
+	// wait for stream to be closed
+	unittest.RequireReturnsBefore(t, senderWG.Wait, 1*time.Second, "could not send messages on time")
+	unittest.RequireReturnsBefore(t, streamCloseWG.Wait, 1*time.Second, "could not close stream at receiver side")
+}
+
+// mockStreamHandlerForMessages creates a stream handler that expects receiving `msgCount` unique messages that match the input regexp.
+// The returned wait group will be unlocked when all messages are completely received and associated streams are closed.
+func mockStreamHandlerForMessages(t *testing.T, ctx context.Context, msgCount int, msgRegexp *regexp.Regexp) (network.StreamHandler, *sync.WaitGroup) {
```

Comment:
:+1: for extracting this piece of logic out into a helper, even if we didn't need to add the wait group